commodore/config.py

```diff
@@ -282,3 +294,119 @@ def _component_is_aliasable(cluster_parameters: dict, component_name: str):
+    """Parse raw dynamic fact value.
+
+    Tries to parse the value as JSON if it starts with the literal `json:`.
+
+    Returns the parsed value or `None` if trying to parse a JSON value results in a
+    decode error."""
+    if raw_value.startswith("json:"):
+        json_val = raw_value.replace("json:", "", 1)
+        # Parse value as JSON if it starts with `json:`, skip value completely
+        # on parse errors.
+        try:
+            v = json.loads(json_val)
+        except json.JSONDecodeError as e:
+            click.secho(
+                f" > Expected value '{json_val}' to be parsable JSON, "
+                + f"but parsing failed with '{e}', skipping",
+                fg="yellow",
+            )
+            return None
+    else:
+        v = raw_value
+    return v
+
+
+def parse_dynamic_facts_from_cli(raw_facts: Iterable[str]) -> dict[str, Any]:
+    """Parse dynamic facts dictionary from strings provided on command line.
+
+    The function expects each raw fact (string) to be of the form `key=value`, where key
+    can contain dots to specify nested keys (e.g. `path.to.key`), and value is parsed as
+    JSON when it starts with the literal `json:`.
+
+    Inputs are processed in order, and subsequent inputs setting the same key will
+    overwrite any existing values. Facts for nested keys will be skipped, if a parent
+    key already exists *and* isn't a dictionary.
+
+    Facts with values marked as JSON which result in a decode error will be skipped.
+
+    Returns a dict with the parsed dynamic facts structure.
+    """
+    facts: dict[str, Any] = {}
+
+    for f in raw_facts:
+        if "=" not in f:
+            click.secho(
+                f" > Ignoring dynamic fact {f} which is not in format key=value",
+                fg="yellow",
+            )
+            continue
+        raw_key, raw_value = f.split("=", maxsplit=1)
+        if not raw_value:
+            click.secho(
+                f"Ignoring malformed dynamic fact {f} with no value. "
+                + "Please specify empty string value as 'json:\"\"'",
+                fg="yellow",
+            )
+            continue
+
+        # Parse value first, so we never add empty dicts for a nested key whose JSON
+        # value turns out to be malformed.
+        value = parse_dynamic_fact_value(raw_value)
+        if value is None:
+            # skip when we failed to parse a value that identified itself as JSON
+            continue
+
+        key, target_fact = parse_sub_key(facts, raw_key)
+        if key is None:
+            # Skip if we found a non-dict entry for a parent of the leaf key or if the
+            # nested key was malformed.
+            continue
+
+        if key in target_fact:
+            click.secho(
+                " > Overwriting dynamic fact "
+                + f"{raw_key}={target_fact[key]} with '{value}'",
+                fg="yellow",
+            )
+        target_fact[key] = value
```

Comment:
Nitpick: I found the `parse_sub_key` function fairly confusing at first as I wasn't expecting it to return a sub dict and was confused how the actual dict was updated. I would probably done something like `set_fact_value(facts, raw_key, value)` and do both key parsing an setting the value in the helper function, but that's just personal preference so :shrug:

Reply:
Probably makes more sense since the diagnostic messages in `parse_sub_key()` don't really make sense if the function was used somewhere else.